device/iot-device-samples/device-reconnection-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/DeviceClientManager.java

```diff
@@ -0,0 +1,146 @@
+        this.connectionStatus = ConnectionStatus.DISCONNECTED;
+        this.client = deviceClient;
+        this.autoReconnectOnDisconnected = autoReconnectOnDisconnected;
+    }
+
+    public void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback callback, Object callbackContext) {
+        this.suppliedConnectionStatusChangeCallback = callback;
+        this.client.registerConnectionStatusChangeCallback(this, callbackContext);
+    }
+
+    public void open() {
+        synchronized (lock) {
+            if(connectionStatus == ConnectionStatus.DISCONNECTED) {
+                connectionStatus = ConnectionStatus.CONNECTING;
+            } else {
+                return;
+            }
+        }
+        doConnect();
+    }
+
+    private void doConnect() {
+        // Device client does not have retry on the initial open() call. Will need to be re-opened by the calling application
+        while (connectionStatus == ConnectionStatus.CONNECTING) {
+            synchronized (lock) {
+                if(connectionStatus == ConnectionStatus.CONNECTING) {
+                    try {
+                        log.debug("[connect] - Opening the device client instance...");
+                        client.open();
+                        connectionStatus = ConnectionStatus.CONNECTED;
+                        break;
+                    }
+                    catch (Exception ex) {
+                        log.error("[connect] - Exception thrown while opening DeviceClient instance: ", ex);
+                    }
+                }
+            }
+
+            try {
+                log.debug("[connect] - Sleeping for 10 secs before attempting another open()");
+                Thread.sleep(SLEEP_TIME_BEFORE_RECONNECTING_IN_SECONDS * 1000);
+            }
+            catch (InterruptedException ex) {
+                log.error("[connect] - Exception in thread sleep: ", ex);
+            }
+        }
+    }
+
+    public void closeNow() {
+        synchronized (lock) {
+            try {
+                log.debug("[disconnect] - Closing the device client instance...");
+                client.closeNow();
+            }
+            catch (IOException e) {
+                log.error("[disconnect] - Exception thrown while closing DeviceClient instance: ", e);
+            } finally {
+                connectionStatus = ConnectionStatus.DISCONNECTED;
+            }
+        }
+
+    }
+
+    @Override
+    public void execute(IotHubConnectionStatus status, IotHubConnectionStatusChangeReason statusChangeReason, Throwable throwable, Object callbackContext) {
+        IotHubConnectionStatusChangeCallback suppliedCallback = this.suppliedConnectionStatusChangeCallback;
+        if (suppliedCallback != null) {
+            log.trace("Invoking the supplied IotHubConnectionStatusChangeCallback implementation.");
+            suppliedCallback.execute(status, statusChangeReason, throwable, callbackContext);
```

Comment:
We need to change the event to `DISCONNECTED _RETRYING` if `autoReconnectOnDisconnected && status == DISCONNECTED && (statusChangeReason == RETRY_EXPIRED || statusChangeReason == COMMUNICATION_ERROR)`